decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java

```diff
@@ -0,0 +1,210 @@
+package openlr.decoder.backtracking;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class BackTrackingGraph<T> {
+    private final List<Slice> slices;
+    private final Map<String, GraphEdge> edges;
+    private int currentSourceSliceIndex = -1;
+    private GraphEdge<T> graphEdge = null;
+    private GraphEdge<T> lastSuccessfulEdge = null;
+
+    public BackTrackingGraph(List<Slice> slices) {
+        this.slices = slices;
+        this.edges = new HashMap<>();
+    }
+
+    private int findNextNodeIndexIfCurrentIsRed(Slice sourceSlice) {
+        if (graphEdge.getSourceNode().isDeadEnd() || !graphEdge.getSourceNode().hasUnmarkedEdges()) {
+            return sourceSlice.getNodePosition(sourceSlice.getNextBestNode());
+        } else {
+            return sourceSlice.getNodePosition(graphEdge.getSourceNode());
+        }
+    }
+
+    private int getNextSourceNodeIndex(Slice sourceSlice) {
+        if (graphEdge == null) {
+            GraphNode sourceNode = sourceSlice.getNextBestNode();
+            return sourceSlice.getNodePosition(sourceNode);
+        } else if (graphEdge.getEdgeColor().equals(EntityColor.RED)) {
+            return findNextNodeIndexIfCurrentIsRed(sourceSlice);
+        } else {
+            return sourceSlice.getNodePosition(graphEdge.getDestinationNode());
+        }
+
+    }
+
+    public boolean isBacktrackingRequired(int currentSliceIndex, GraphEdge<T> graphEdge) {
```

Comment:
Many of the methods are public but I'm wondering if this needs to be.  From what you showed me earlier, the client just needs to call 1 or 2 methods.

Reply:
fixed